test/libsolidity/semanticTests/externalContracts/strings.sol

```diff
@@ -51,26 +51,26 @@ contract test {
 // compileViaYul: also
 // ----
 // constructor()
-// gas irOptimized: 912777
+// gas irOptimized: 852919
 // gas legacy: 1188228
 // gas legacyOptimized: 771634
 // toSlice(string): 0x20, 11, "hello world" -> 11, 0xa0
-// gas irOptimized: 22877
+// gas irOptimized: 22628
 // gas legacy: 23190
 // gas legacyOptimized: 22508
 // roundtrip(string): 0x20, 11, "hello world" -> 0x20, 11, "hello world"
-// gas irOptimized: 23676
+// gas irOptimized: 23506
 // gas legacy: 23820
 // gas legacyOptimized: 23123
 // utf8len(string): 0x20, 16, "\xf0\x9f\x98\x83\xf0\x9f\x98\x83\xf0\x9f\x98\x83\xf0\x9f\x98\x83" -> 4 # Input: "😃😃😃😃" #
-// gas irOptimized: 24779
+// gas irOptimized: 24239
 // gas legacy: 25716
 // gas legacyOptimized: 24115
 // multiconcat(string,uint256): 0x40, 3, 11, "hello world" -> 0x20, 0x58, 0x68656c6c6f20776f726c6468656c6c6f20776f726c6468656c6c6f20776f726c, 0x6468656c6c6f20776f726c6468656c6c6f20776f726c6468656c6c6f20776f72, 49027192869463622675296414541903001712009715982962058146354235762728281047040 # concatenating 3 times #
-// gas irOptimized: 29729
+// gas irOptimized: 28808
 // gas legacy: 31621
 // gas legacyOptimized: 27914
 // benchmark(string,bytes32): 0x40, 0x0842021, 8, "solidity" -> 0x2020
-// gas irOptimized: 2903627
+// gas irOptimized: 1872927
```

Comment:
Beating legacy quite a bit!

Reply:
Code is larger :(

Reply:
I do excessively generate labels for each and every block at the moment - that's at least part of the code size... But yeah, it generally seems to increase.

Reply:
Are those labels also used, or just generated?

Because in libevmasm we have an optimiser step to remove unused labels, so with #10286 that should reduce it. Alternatively you could maintain a map of usage and drop unused labels in codegen.

Reply:
No, that's what I meant: I generate quite a few labels that *won't* be used in the end and I think the pipeline currently doesn't remove it.
And yes: I partly do that, because the libevmasm step that removes unused labels is among the simplest ones, so I thought, it might be easier to just generate labels everywhere and remove them afterwards. I can also try to only generate the needed ones in the first place, but it's not so far up on my list :-). Although it'd be interesting to know, if the larger code size is *only* due to that or not...

Reply:
Hm, actually, looking at it again, it's only *some* cases, not *all the time* that I generate unused labels, so I guess the increased code size is primarily due to some other reason.

Reply:
Still both running the ``BlockDeduplicator`` and the ``UnusedJumpdestRemover`` on the resulting assembly would be interesting.

Reply:
I'm sure you can easily rebase #10286 on top of this to try?

Reply:
With full #10286 for this particular test I get
```
  // constructor()
  // gas irOptimized: 761570 [-103698 (-11%)]
  // gas legacy: 1188228
  // gas legacyOptimized: 771634
```
But then again, I also still get this :-D:
```
// benchmark(string,bytes32): 0x40, 0x0842021, 8, "solidity" -> 0x2020
// gas irOptimized: 1678601 [-194351 (-10%)]
 ```
Let me see what happens with only the deduplicator and jumpdest remover...

Reply:
(For the deposit contract full #10286 is an additional 2% less gas on runtime code and 4% less gas for the constructor)

Reply:
Constructor gas cost for ``strings.sol`` reduces by:
- a total of 1% with only jumpdest remover and block deduplicator
- a total of 3% with jumpdest remover, block deduplicator and constant optimiser
- a total of 8% with jumpdest remover, block deduplicator, constant optimiser and peephole optimiser
- a total of 9% with jumpdest remover, block deduplicator, constant optimiser, peephole optimiser and CSE
- a total of 11%  with full evmasm stack, i.e. jumpdest remover, block deduplicator, constant optimiser, peephole optimiser, CSE and inliner

The benefit for ``benchmark`` is
- a total of 0% with only jumpdest remover and block deduplicator
- a total of 0% with jumpdest remover, block deduplicator and constant optimiser
- a total of 3% with jumpdest remover, block deduplicator, constant optimiser and peephole optimiser
- a total of 5% with jumpdest remover, block deduplicator, constant optimiser, peephole optimiser and CSE
- a total of 10%  with full evmasm stack, i.e. jumpdest remover, block deduplicator, constant optimiser, peephole optimiser, 

Surprisingly, even though by design the inliner is intended to run only in combination with both CSE and peephole optimiser, I get the following with everything *except* CSE (that is with jumpdest remover, block deduplicator, constant optimiser, peephole optimiser and inliner, which are all comparably simple steps):
```
  // constructor()
  // gas irOptimized: 770588 [-94680 (-10%)]
  // benchmark(string,bytes32): 0x40, 0x0842021, 8, "solidity" -> 0x2020
  // gas irOptimized: 1703469 [-169483 (-9%)]
  // gas legacy: 4381235
  // gas legacyOptimized: 2317529
```
Maybe the constant optimiser is not *that* simple. Dropping that as well (i.e. jumpdest remover, block deduplicator, peephole optimiser and inliner), I get:
```
  // constructor()
  // gas irOptimized: 781965 [-83303 (-9%)]
  // gas legacy: 1188228
  // gas legacyOptimized: 771634
  // benchmark(string,bytes32): 0x40, 0x0842021, 8, "solidity" -> 0x2020
  // gas irOptimized: 1703463 [-169489 (-9%)]
  // gas legacy: 4381235
  // gas legacyOptimized: 2317529
```

Reply:
I'm a bit surprised that the peephole optimiser seems to have the greatest effect. Maybe I'm generating some junk somewhere :-). Still all in all I'd say this is looking quite good.

Reply:
Just for the record: this is the gas savings summarized that I get from this PR plus #10286 (but without CSE):
<details><summary>Click for a table of gas differences</summary>

| file name | IR-Opti | Legacy-Opti | Legacy|
|-----------|---------|-------------|-------|
| interface_inheritance_conversions.sol |    -19.352% |      0.000% |      0.000% |
| events/event_indexed_string.sol |     -8.438% |      0.000% |      0.000% |
| events/event_indexed_mixed.sol |     -0.038% |      0.000% |      0.000% |
| events/event_dynamic_nested_array_storage_v2.sol |     -1.179% |      0.000% |      0.000% |
| events/event_dynamic_array_storage_v2.sol |     -1.247% |      0.000% |      0.000% |
| events/event_dynamic_array_storage.sol |     -1.247% |      0.000% |      0.000% |
| functionCall/failed_create.sol |    -15.806% |      0.000% |      0.000% |
| functionCall/mapping_array_internal_argument.sol |     -0.317% |      0.000% |      0.000% |
| functionCall/inheritance/call_unimplemented_base.sol |      0.000% |      0.000% |      0.000% |
| externalContracts/snark.sol |    -21.060% |      0.000% |      0.000% |
| externalContracts/ramanujan_pi.sol |    -19.495% |      0.000% |      0.000% |
| externalContracts/prbmath_unsigned.sol |     -4.081% |      0.000% |      0.000% |
| externalContracts/deposit_contract.sol |     -9.963% |      0.000% |      0.000% |
| externalContracts/prbmath_signed.sol |     -5.205% |      0.000% |      0.000% |
| externalContracts/strings.sol |    -34.330% |      0.000% |      0.000% |
| constructor/arrays_in_constructors.sol |     -6.942% |      0.000% |      0.000% |
| constructor/no_callvalue_check.sol |     -0.585% |      0.000% |      0.000% |
| constructor/bytes_in_constructors_packer.sol |     -8.251% |      0.000% |      0.000% |
| abiencodedecode/abi_decode_simple_storage.sol |      0.071% |      0.000% |      0.000% |
| inlineAssembly/keccak_yul_optimization.sol |     -0.829% |      0.000% |      0.000% |
| inheritance/inherited_function_calldata_calldata_interface.sol |    -10.202% |      0.000% |      0.000% |
| inheritance/address_overload_resolution.sol |    -11.506% |      0.000% |      0.000% |
| inheritance/inherited_function_calldata_memory_interface.sol |     -4.349% |      0.000% |      0.000% |
| functionTypes/store_function.sol |     -1.937% |      0.000% |      0.000% |
| storage/packed_storage_structs_bytes.sol |     -1.217% |      0.000% |      0.000% |
| abiEncoderV2/abi_encode_calldata_slice.sol |    -19.523% |      0.000% |      0.000% |
| abiEncoderV2/storage_array_encoding.sol |      0.002% |      0.000% |      0.000% |
| abiEncoderV2/abi_encode_v2_in_modifier_used_in_v1_contract.sol |     -4.427% |      0.000% |      0.000% |
| abiEncoderV2/abi_encode_v2_in_function_inherited_in_v1_contract.sol |     -0.364% |      0.000% |      0.000% |
| abiEncoderV2/calldata_array.sol |    -42.379% |      0.000% |      0.000% |
| abiEncoderV2/abi_encode_v2.sol |     -0.356% |      0.000% |      0.000% |
| abiEncoderV1/abi_encode_calldata_slice.sol |    -19.523% |      0.000% |      0.000% |
| abiEncoderV1/abi_decode_v2_storage.sol |      0.378% |      0.000% |      0.000% |
| abiEncoderV1/struct/struct_storage_ptr.sol |     -0.731% |      0.000% |      0.000% |
| viaYul/array_storage_length_access.sol |    -10.797% |      0.000% |      0.000% |
| viaYul/array_storage_index_boundary_test.sol |     -9.764% |      0.000% |      0.000% |
| viaYul/array_memory_index_access.sol |    -10.885% |      0.000% |      0.000% |
| viaYul/array_storage_index_access.sol |     -3.525% |      0.000% |      0.000% |
| viaYul/array_storage_push_empty_length_address.sol |    -10.098% |      0.000% |      0.000% |
| viaYul/array_storage_push_empty.sol |     -1.666% |      0.000% |      0.000% |
| viaYul/array_storage_push_pop.sol |     -3.555% |      0.000% |      0.000% |
| viaYul/simple.sol |      0.000% |      0.000% |      0.000% |
| viaYul/array_storage_index_zeroed_test.sol |     -3.923% |      0.000% |      0.000% |
| salted_create/salted_create.sol |     -0.000% |      0.000% |      0.000% |
| salted_create/salted_create_with_value.sol |     -7.551% |      0.000% |      0.000% |
| array/function_array_cross_calls.sol |      3.574% |      0.000% |      0.000% |
| array/reusing_memory.sol |     -1.877% |      0.000% |      0.000% |
| array/byte_array_transitional_2.sol |     -7.457% |      0.000% |      0.000% |
| array/bytes_length_member.sol |     -0.061% |      0.000% |      0.000% |
| array/dynamic_multi_array_cleanup.sol |     -6.352% |      0.000% |      0.000% |
| array/byte_array_storage_layout.sol |     -2.520% |      0.000% |      0.000% |
| array/dynamic_array_cleanup.sol |     -0.206% |      0.000% |      0.000% |
| array/fixed_arrays_as_return_type.sol |    -17.255% |      0.000% |      0.000% |
| array/dynamic_arrays_in_storage.sol |     -4.107% |      0.000% |      0.000% |
| array/arrays_complex_from_and_to_storage.sol |     -0.947% |      0.000% |      0.000% |
| array/fixed_array_cleanup.sol |     -0.074% |      0.000% |      0.000% |
| array/create_memory_array.sol |     -3.105% |      0.000% |      0.000% |
| array/delete/delete_storage_array_packed.sol |     -1.155% |      0.000% |      0.000% |
| array/delete/bytes_delete_element.sol |    -15.218% |      0.000% |      0.000% |
| array/push/nested_bytes_push.sol |     -0.485% |      0.000% |      0.000% |
| array/push/array_push_packed_array.sol |     -1.739% |      0.000% |      0.000% |
| array/push/array_push_struct_from_calldata.sol |     -0.633% |      0.000% |      0.000% |
| array/push/array_push_nested_from_calldata.sol |     -1.215% |      0.000% |      0.000% |
| array/push/byte_array_push_transition.sol |     -9.736% |      0.000% |      0.000% |
| array/push/push_no_args_2d.sol |     -2.133% |      0.000% |      0.000% |
| array/push/array_push_struct.sol |     -0.886% |      0.000% |      0.000% |
| array/push/push_no_args_bytes.sol |     -4.476% |      0.000% |      0.000% |
| array/push/array_push.sol |     -0.367% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_different_base_nested.sol |     -0.484% |      0.000% |      0.000% |
| array/copying/array_copy_storage_to_memory_nested.sol |     -0.665% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_static_static.sol |     -0.175% |      0.000% |      0.000% |
| array/copying/copy_function_storage_array.sol |     -0.484% |      0.000% |      0.000% |
| array/copying/array_copy_different_packing.sol |     -1.857% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_static_dynamic.sol |      0.001% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_dynamic_dynamic.sol |     -0.191% |      0.000% |      0.000% |
| array/copying/array_copy_clear_storage_packed.sol |     -1.406% |      0.000% |      0.000% |
| array/copying/array_copy_target_simple_2.sol |     -0.540% |      0.000% |      0.000% |
| array/copying/array_nested_calldata_to_storage.sol |     -0.749% |      0.000% |      0.000% |
| array/copying/memory_dyn_2d_bytes_to_storage.sol |     -9.129% |      0.000% |      0.000% |
| array/copying/array_storage_multi_items_per_slot.sol |     -1.040% |      0.000% |      0.000% |
| array/copying/array_nested_memory_to_storage.sol |     -0.740% |      0.000% |      0.000% |
| array/copying/array_copy_including_array.sol |     -0.557% |      0.000% |      0.000% |
| array/copying/array_copy_target_simple.sol |     -0.790% |      0.000% |      0.000% |
| array/copying/array_copy_cleanup_uint128.sol |     -1.343% |      0.000% |      0.000% |
| array/copying/copy_byte_array_in_struct_to_storage.sol |      0.148% |      0.000% |      0.000% |
| array/copying/array_copy_target_leftover.sol |     -9.364% |      0.000% |      0.000% |
| array/copying/calldata_array_dynamic_to_storage.sol |     -0.216% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_struct.sol |     -0.456% |      0.000% |      0.000% |
| array/copying/bytes_inside_mappings.sol |     -0.026% |      0.000% |      0.000% |
| array/copying/array_of_struct_memory_to_storage.sol |     -0.867% |      0.000% |      0.000% |
| array/copying/copy_byte_array_to_storage.sol |     -1.512% |      0.000% |      0.000% |
| array/copying/array_copy_calldata_storage.sol |      0.060% |      0.000% |      0.000% |
| array/copying/copy_removes_bytes_data.sol |      0.016% |      0.000% |      0.000% |
| array/copying/array_copy_target_leftover2.sol |     -3.367% |      0.000% |      0.000% |
| array/copying/array_copy_nested_array.sol |     -0.088% |      0.000% |      0.000% |
| array/copying/array_of_struct_calldata_to_storage.sol |     -0.548% |      0.000% |      0.000% |
| array/copying/storage_memory_nested.sol |     -1.518% |      0.000% |      0.000% |
| array/copying/storage_memory_packed_dyn.sol |     -4.031% |      0.000% |      0.000% |
| array/copying/array_of_structs_containing_arrays_memory_to_storage.sol |     -1.120% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_dyn_dyn.sol |      0.202% |      0.000% |      0.000% |
| array/copying/storage_memory_nested_bytes.sol |      0.158% |      0.000% |      0.000% |
| array/copying/storage_memory_nested_struct.sol |     -0.529% |      0.000% |      0.000% |
| array/copying/array_copy_storage_storage_different_base.sol |     -0.449% |      0.000% |      0.000% |
| array/copying/array_copy_clear_storage.sol |     -0.642% |      0.000% |      0.000% |
| array/copying/array_of_structs_containing_arrays_calldata_to_storage.sol |     -3.323% |      0.000% |      0.000% |
| array/copying/storage_memory_nested_from_pointer.sol |     -1.526% |      0.000% |      0.000% |
| array/copying/arrays_from_and_to_storage.sol |     -0.444% |      0.000% |      0.000% |
| array/copying/bytes_storage_to_storage.sol |    -17.234% |      0.000% |      0.000% |
| array/copying/array_copy_cleanup_uint40.sol |     -4.619% |      0.000% |      0.000% |
| array/pop/array_pop_uint24_transition.sol |     -3.287% |      0.000% |      0.000% |
| array/pop/byte_array_pop_masking_long.sol |     -0.404% |      0.000% |      0.000% |
| array/pop/array_pop_uint16_transition.sol |     -4.594% |      0.000% |      0.000% |
| array/pop/byte_array_pop_copy_long.sol |     -1.706% |      0.000% |      0.000% |
| array/pop/byte_array_pop_long_storage_empty_garbage_ref.sol |     -5.094% |      0.000% |      0.000% |
| array/pop/byte_array_pop_long_storage_empty.sol |    -17.718% |      0.000% |      0.000% |
| array/pop/array_pop_array_transition.sol |     -0.293% |      0.000% |      0.000% |
| immutable/multi_creation.sol |     -1.753% |      0.000% |      0.000% |
| structs/struct_copy_via_local.sol |     -0.504% |      0.000% |      0.000% |
| structs/struct_delete_storage_nested_small.sol |     -0.310% |      0.000% |      0.000% |
| structs/struct_delete_storage_with_array.sol |     -0.453% |      0.000% |      0.000% |
| structs/structs.sol |     -0.205% |      0.000% |      0.000% |
| structs/memory_structs_nested_load.sol |     -1.364% |      0.000% |      0.000% |
| structs/struct_copy.sol |     -0.144% |      0.000% |      0.000% |
| structs/struct_delete_storage_with_arrays_small.sol |     -0.634% |      0.000% |      0.000% |
| structs/struct_containing_bytes_copy_and_delete.sol |     -0.120% |      0.000% |      0.000% |
| structs/struct_memory_to_storage_function_ptr.sol |     -0.950% |      0.000% |      0.000% |
| structs/conversion/recursive_storage_memory.sol |     -0.419% |      0.000% |      0.000% |
| structs/calldata/calldata_struct_with_nested_array_to_storage.sol |     -0.227% |      0.000% |      0.000% |
| various/staticcall_for_view_and_pure.sol |     -0.000% |      0.000% |      0.000% |
| various/destructuring_assignment.sol |     -0.990% |      0.000% |      0.000% |
| various/swap_in_storage_overwrite.sol |     -0.062% |      0.000% |      0.000% |
| various/skip_dynamic_types_for_structs.sol |     -0.131% |      0.000% |      0.000% |

Reply:
I'll leave it at that analysis-wise, though, and get back to simplifying this beast.

Reply:
Actually, one more. This is the difference between #10286 rebased on develop (i.e. the baseline is the current evm code transform plus running the optimizer of libevmasm) compared to the new code generator in this PR plus the libevmasm optimiser (i.e. #10286 merged into this branch).

Notably, there are some cases in which this is unfavorable. But fortunately, in general, this still looks good. Among the external tests:

| file name | IR-Opti |
|-----------|---------|
| externalContracts/snark.sol |    -13.393%
| externalContracts/ramanujan_pi.sol |     -9.108% |
| externalContracts/prbmath_unsigned.sol |      1.685% |
| externalContracts/deposit_contract.sol |     -4.354% | 
| externalContracts/prbmath_signed.sol |      2.075% |
| externalContracts/strings.sol |    -18.420% | 


And in all details:
<details><summary>Click for a table of gas differences</summary>

| file name | IR-Opti | Legacy-Opti | Legacy|
|-----------|---------|-------------|-------|
| interface_inheritance_conversions.sol |     -5.804% |
| events/event_indexed_string.sol |     -4.938% |
| events/event_indexed_mixed.sol |     -0.038% |
| events/event_dynamic_nested_array_storage_v2.sol |     -1.113% |
| events/event_dynamic_array_storage_v2.sol |     -1.036% |
| events/event_dynamic_array_storage.sol |     -1.036% |
| functionCall/failed_create.sol |     -4.068% |
| functionCall/mapping_array_internal_argument.sol |     -0.217% |
| functionCall/inheritance/call_unimplemented_base.sol |      0.000% |
| externalContracts/snark.sol |    -13.393% |
| externalContracts/ramanujan_pi.sol |     -9.108% |
| externalContracts/prbmath_unsigned.sol |      1.685% |
| externalContracts/deposit_contract.sol |     -4.354% |
| externalContracts/prbmath_signed.sol |      2.075% |
| externalContracts/strings.sol |    -18.420% |
| constructor/arrays_in_constructors.sol |     -1.960% |
| constructor/no_callvalue_check.sol |     -0.063% |
| constructor/bytes_in_constructors_packer.sol |     -3.849% |
| abiencodedecode/abi_decode_simple_storage.sol |     -0.179% |
| inlineAssembly/keccak_yul_optimization.sol |     -0.241% |
| inheritance/inherited_function_calldata_calldata_interface.sol |     -5.470% |
| inheritance/address_overload_resolution.sol |     -4.107% |
| inheritance/inherited_function_calldata_memory_interface.sol |      1.951% |
| functionTypes/store_function.sol |     -0.904% |
| storage/packed_storage_structs_bytes.sol |     -1.282% |
| abiEncoderV2/abi_encode_calldata_slice.sol |     -4.090% |
| abiEncoderV2/storage_array_encoding.sol |     -0.030% |
| abiEncoderV2/abi_encode_v2_in_modifier_used_in_v1_contract.sol |     -0.857% |
| abiEncoderV2/abi_encode_v2_in_function_inherited_in_v1_contract.sol |      6.920% |
| abiEncoderV2/calldata_array.sol |    -26.856% |
| abiEncoderV2/abi_encode_v2.sol |     -0.344% |
| abiEncoderV1/abi_encode_calldata_slice.sol |     -4.090% |
| abiEncoderV1/abi_decode_v2_storage.sol |     -0.153% |
| abiEncoderV1/struct/struct_storage_ptr.sol |     -0.606% |
| viaYul/array_storage_length_access.sol |     -0.309% |
| viaYul/array_storage_index_boundary_test.sol |     -1.942% |
| viaYul/array_memory_index_access.sol |     -1.858% |
| viaYul/array_storage_index_access.sol |     -2.256% |
| viaYul/array_storage_push_empty_length_address.sol |     -1.216% |
| viaYul/array_storage_push_empty.sol |     -1.626% |
| viaYul/array_storage_push_pop.sol |     -0.851% |
| viaYul/simple.sol |      0.000% |
| viaYul/array_storage_index_zeroed_test.sol |     -2.383% |
| salted_create/salted_create.sol |     -0.000% |
| salted_create/salted_create_with_value.sol |      0.870% |
| array/function_array_cross_calls.sol |     10.685% |
| array/reusing_memory.sol |     -0.767% |
| array/byte_array_transitional_2.sol |     -3.506% |
| array/bytes_length_member.sol |     -0.022% |
| array/dynamic_multi_array_cleanup.sol |     -5.312% |
| array/byte_array_storage_layout.sol |     -1.378% |
| array/dynamic_array_cleanup.sol |     -0.083% |
| array/fixed_arrays_as_return_type.sol |     -9.152% |
| array/dynamic_arrays_in_storage.sol |     -0.929% |
| array/arrays_complex_from_and_to_storage.sol |     -0.056% |
| array/fixed_array_cleanup.sol |     -0.033% |
| array/create_memory_array.sol |      3.198% |
| array/delete/delete_storage_array_packed.sol |     -1.318% |
| array/delete/bytes_delete_element.sol |    -10.381% |
| array/push/nested_bytes_push.sol |     -0.385% |
| array/push/array_push_packed_array.sol |     -1.869% |
| array/push/array_push_struct_from_calldata.sol |     -0.290% |
| array/push/array_push_nested_from_calldata.sol |     -0.969% |
| array/push/byte_array_push_transition.sol |     -4.552% |
| array/push/push_no_args_2d.sol |      0.081% |
| array/push/array_push_struct.sol |     -0.174% |
| array/push/push_no_args_bytes.sol |     -3.609% |
| array/push/array_push.sol |     -0.385% |
| array/copying/array_copy_storage_storage_different_base_nested.sol |     -0.279% |
| array/copying/array_copy_storage_to_memory_nested.sol |     -0.581% |
| array/copying/array_copy_storage_storage_static_static.sol |      0.045% |
| array/copying/copy_function_storage_array.sol |      0.447% |
| array/copying/array_copy_different_packing.sol |     -1.281% |
| array/copying/array_copy_storage_storage_static_dynamic.sol |      0.056% |
| array/copying/array_copy_storage_storage_dynamic_dynamic.sol |     -0.279% |
| array/copying/array_copy_clear_storage_packed.sol |     -1.309% |
| array/copying/array_copy_target_simple_2.sol |     -0.278% |
| array/copying/array_nested_calldata_to_storage.sol |     -0.497% |
| array/copying/memory_dyn_2d_bytes_to_storage.sol |     -7.035% |
| array/copying/array_storage_multi_items_per_slot.sol |     -0.504% |
| array/copying/array_nested_memory_to_storage.sol |     -0.526% |
| array/copying/array_copy_including_array.sol |     -0.220% |
| array/copying/array_copy_target_simple.sol |     -0.437% |
| array/copying/array_copy_cleanup_uint128.sol |     -1.314% |
| array/copying/copy_byte_array_in_struct_to_storage.sol |     -0.039% |
| array/copying/array_copy_target_leftover.sol |     -3.694% |
| array/copying/calldata_array_dynamic_to_storage.sol |     -0.163% |
| array/copying/array_copy_storage_storage_struct.sol |     -0.059% |
| array/copying/bytes_inside_mappings.sol |     -0.051% |
| array/copying/array_of_struct_memory_to_storage.sol |     -0.675% |
| array/copying/copy_byte_array_to_storage.sol |     -1.205% |
| array/copying/array_copy_calldata_storage.sol |      0.047% |
| array/copying/copy_removes_bytes_data.sol |      0.002% |
| array/copying/array_copy_target_leftover2.sol |     -2.104% |
| array/copying/array_copy_nested_array.sol |     -0.091% |
| array/copying/array_of_struct_calldata_to_storage.sol |     -0.418% |
| array/copying/storage_memory_nested.sol |     -0.558% |
| array/copying/storage_memory_packed_dyn.sol |     -3.364% |
| array/copying/array_of_structs_containing_arrays_memory_to_storage.sol |     -0.518% |
| array/copying/array_copy_storage_storage_dyn_dyn.sol |     -0.061% |
| array/copying/storage_memory_nested_bytes.sol |      0.064% |
| array/copying/storage_memory_nested_struct.sol |     -0.145% |
| array/copying/array_copy_storage_storage_different_base.sol |     -0.311% |
| array/copying/array_copy_clear_storage.sol |     -0.473% |
| array/copying/array_of_structs_containing_arrays_calldata_to_storage.sol |     -2.824% |
| array/copying/storage_memory_nested_from_pointer.sol |     -0.550% |
| array/copying/arrays_from_and_to_storage.sol |      1.127% |
| array/copying/bytes_storage_to_storage.sol |    -10.679% |
| array/copying/array_copy_cleanup_uint40.sol |     -4.404% |
| array/pop/array_pop_uint24_transition.sol |     -1.716% |
| array/pop/byte_array_pop_masking_long.sol |      0.416% |
| array/pop/array_pop_uint16_transition.sol |     -2.836% |
| array/pop/byte_array_pop_copy_long.sol |     -0.640% |
| array/pop/byte_array_pop_long_storage_empty_garbage_ref.sol |     -2.508% |
| array/pop/byte_array_pop_long_storage_empty.sol |    -11.879% |
| array/pop/array_pop_array_transition.sol |     -0.093% |
| immutable/multi_creation.sol |     -0.705% |
| structs/struct_copy_via_local.sol |     -0.433% |
| structs/struct_delete_storage_nested_small.sol |     -0.146% |
| structs/struct_delete_storage_with_array.sol |     -0.448% |
| structs/structs.sol |      0.063% |
| structs/memory_structs_nested_load.sol |     -0.882% |
| structs/struct_copy.sol |      0.001% |
| structs/struct_delete_storage_with_arrays_small.sol |     -0.447% |
| structs/struct_containing_bytes_copy_and_delete.sol |     -0.123% |
| structs/struct_memory_to_storage_function_ptr.sol |     -0.397% |
| structs/conversion/recursive_storage_memory.sol |      1.194% |
| structs/calldata/calldata_struct_with_nested_array_to_storage.sol |      0.093% |
| various/staticcall_for_view_and_pure.sol |      0.000% |
| various/destructuring_assignment.sol |     -0.924% |
| various/swap_in_storage_overwrite.sol |     -0.044% |
| various/skip_dynamic_types_for_structs.sol |     -0.132% |

</details>

Given that there's room to improve this PR at several places (the stack shuffling is not yet optimal and in particular when the stack limit is reached, there is better workarounds than blindly dupping deepest first and swapping, etc.), and still a lot of the cases already get better compared to old-code-transform plus libevmasm-optimizer, I'm rather confident in this whole thing. Also for the ``prbmath``, note the diff between develop + libevmasm and this PR + libevmasm split into constructor and benchmark:

```diff
-- a/test/libsolidity/semanticTests/externalContracts/prbmath_signed.sol
+++ b/test/libsolidity/semanticTests/externalContracts/prbmath_signed.sol
 // constructor()
-// gas irOptimized: 1947094
+// gas irOptimized: 1992997
...
 // benchmark(int256): 3141592653589793238 -> 998882724338592125, 1000000000000000000, 1000000000000000000
-// gas irOptimized: 37899
+// gas irOptimized: 33540
```

```diff
--- a/test/libsolidity/semanticTests/externalContracts/prbmath_unsigned.sol
+++ b/test/libsolidity/semanticTests/externalContracts/prbmath_unsigned.sol
 // constructor()
-// gas irOptimized: 1770739
+// gas irOptimized: 1806819
...
 // benchmark(uint256): 3141592653589793238 -> 998882724338592125, 1000000000000000000, 1000000000000000000
-// gas irOptimized: 36587
+// gas irOptimized: 34644
```

So runtime still improves, while code size deterioates.